Unable to compile `duckdb-sys` on Windows

We were setting up Windows builds in GitHub Actions for our app and noticed that the `duckdb-sys v0.4.0` crate's build script is unable to compile DuckDB from source.

There are a lot of suspicious warnings, and I've had to attach the output as a file because GitHub has a  65536 character limit on comments, but I think this is the thing that actually fails the build:

```
fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj
```

[(Link to full output from the build script)](https://github.com/wangfenjin/duckdb-rs/files/9007121/stderr.txt)

I notice you've actually commented out the Windows build for this repo's CI. Does that mean this is a known issue?

https://github.com/wangfenjin/duckdb-rs/blob/30d8e3e637f8a3f9b96ca6de9e2275bc69a5bbae/.github/workflows/rust.yaml#L15-L17

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -44,19 +44,19 @@ jobs:
           tag: "v0.4.0"
           fileName: ${{ matrix.duckdb }}
           out-file-path: .
-      - name: Extract duckdb
+      - name: Linux extract duckdb
         if: matrix.os == 'ubuntu-latest'
         uses: ihiroky/extract-action@v1
         with:
           file_path: ${{ github.workspace }}/${{ matrix.duckdb }}
           extract_dir: libduckdb
-      - name: Extract duckdb
+      - name: Windows extract duckdb
         if: matrix.os == 'windows-latest'
-        uses: ihiroky/extract-action@v1
+        uses: extract-7z-action@v1
         with:
           # file_path: ${{ github.workspace }}\${{ matrix.duckdb }} # not work
-          file_path: D:\a\duckdb-rs\duckdb-rs\${{ matrix.duckdb }}
-          extract_dir: ${{ github.workspace }}/libduckdb
+          pathSource: D:\a\duckdb-rs\duckdb-rs\${{ matrix.duckdb }}
+          pathTarget: ${{ github.workspace }}/libduckdb
 
       # - run: cargo clippy --all-targets --workspace --features bundled --features modern-full -- -D warnings -A clippy::redundant-closure
       - run: cargo clippy --all-targets --workspace --features buildtime_bindgen --features modern-full -- -D warnings -A clippy::redundant-closure

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -60,12 +60,21 @@ jobs:
 
       # - run: cargo clippy --all-targets --workspace --features bundled --features modern-full -- -D warnings -A clippy::redundant-closure
       - run: cargo clippy --all-targets --workspace --features buildtime_bindgen --features modern-full -- -D warnings -A clippy::redundant-closure
+        if: matrix.os == 'ubuntu-latest'
         name: run cargo clippy
         env:
           DUCKDB_LIB_DIR: ${{ github.workspace }}/libduckdb
           DUCKDB_INCLUDE_DIR: ${{ github.workspace }}/libduckdb
           LD_LIBRARY_PATH: ${{ github.workspace }}/libduckdb
+      - name: Run cargo-test
+        if: matrix.os == 'windows-latest'
+        run: cargo test --features "buildtime_bindgen" --features "modern-full"
+        env:
+          DUCKDB_LIB_DIR: ${{ github.workspace }}/libduckdb
+          DUCKDB_INCLUDE_DIR: ${{ github.workspace }}/libduckdb
+          LD_LIBRARY_PATH: ${{ github.workspace }}/libduckdb
       - name: Run cargo-tarpaulin
+        if: matrix.os == 'ubuntu-latest'
         uses: actions-rs/tarpaulin@v0.1
         with:
           # Intentionally omit time feature until we're on time 0.3, at which

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -33,10 +33,6 @@ jobs:
           rust-version: stable${{ matrix.host }}
           targets: ${{ matrix.target }}
           components: 'rustfmt, clippy'
-
-      - run: cargo fmt --all -- --check
-        if: matrix.os == 'ubuntu-latest'
-      
       # download libduckdb
       - uses: robinraju/release-downloader@v1.4
         name: Download duckdb
@@ -45,21 +41,16 @@ jobs:
           tag: "v0.4.0"
           fileName: ${{ matrix.duckdb }}
           out-file-path: .
+
+      # For Linux
       - name: Linux extract duckdb
         if: matrix.os == 'ubuntu-latest'
         uses: ihiroky/extract-action@v1
         with:
           file_path: ${{ github.workspace }}/${{ matrix.duckdb }}
           extract_dir: libduckdb
-      - name: Windows extract duckdb
-        if: matrix.os == 'windows-latest'
-        uses: DuckSoft/extract-7z-action@v1.0
-        with:
-          # file_path: ${{ github.workspace }}\${{ matrix.duckdb }} # not work
-          pathSource: D:\a\duckdb-rs\duckdb-rs\${{ matrix.duckdb }}
-          pathTarget: ${{ github.workspace }}/libduckdb
-
-      # - run: cargo clippy --all-targets --workspace --features bundled --features modern-full -- -D warnings -A clippy::redundant-closure
+      - run: cargo fmt --all -- --check
+        if: matrix.os == 'ubuntu-latest'
       - run: cargo clippy --all-targets --workspace --features buildtime_bindgen --features modern-full -- -D warnings -A clippy::redundant-closure
         if: matrix.os == 'ubuntu-latest'
         name: run cargo clippy
@@ -67,6 +58,29 @@ jobs:
           DUCKDB_LIB_DIR: ${{ github.workspace }}/libduckdb
           DUCKDB_INCLUDE_DIR: ${{ github.workspace }}/libduckdb
           LD_LIBRARY_PATH: ${{ github.workspace }}/libduckdb
+      - name: Run cargo-tarpaulin
+        if: matrix.os == 'ubuntu-latest'
+        uses: actions-rs/tarpaulin@v0.1
+        with:
+          # Intentionally omit time feature until we're on time 0.3, at which
+          # point it should be added to `bundled-full`.
+          args: '--features "buildtime_bindgen" --features "modern-full" --avoid-cfg-tarpaulin' # TODO restore to normal (https://github.com/xd009642/tarpaulin/issues/756#issuecomment-838769320)
+        env:
+          DUCKDB_LIB_DIR: ${{ github.workspace }}/libduckdb
+          DUCKDB_INCLUDE_DIR: ${{ github.workspace }}/libduckdb
+          LD_LIBRARY_PATH: ${{ github.workspace }}/libduckdb
+      - name: Upload to codecov.io
+        if: matrix.os == 'ubuntu-latest'
+        uses: codecov/codecov-action@v1
+
+      # For windows
+      - name: Windows extract duckdb
+        if: matrix.os == 'windows-latest'
+        uses: DuckSoft/extract-7z-action@v1.0
+        with:
+          pathSource: D:\a\duckdb-rs\duckdb-rs\${{ matrix.duckdb }}
+          pathTarget: ${{ github.workspace }}/libduckdb
+
       - name: Add path to PATH environment variable
         if: matrix.os == 'windows-latest'
         uses: myci-actions/export-env-var-powershell@1
@@ -79,23 +93,20 @@ jobs:
         env:
           DUCKDB_LIB_DIR: ${{ github.workspace }}/libduckdb
           DUCKDB_INCLUDE_DIR: ${{ github.workspace }}/libduckdb
-      - name: Run cargo-tarpaulin
-        if: matrix.os == 'ubuntu-latest'
-        uses: actions-rs/tarpaulin@v0.1
-        with:
-          # Intentionally omit time feature until we're on time 0.3, at which
-          # point it should be added to `bundled-full`.
-          args: '--features "buildtime_bindgen" --features "modern-full" --avoid-cfg-tarpaulin' # TODO restore to normal (https://github.com/xd009642/tarpaulin/issues/756#issuecomment-838769320)
-          # args: '--features "bundled" --features "modern-full" --avoid-cfg-tarpaulin' # TODO restore to normal (https://github.com/xd009642/tarpaulin/issues/756#issuecomment-838769320)
-        env:
-          DUCKDB_LIB_DIR: ${{ github.workspace }}/libduckdb
-          DUCKDB_INCLUDE_DIR: ${{ github.workspace }}/libduckdb
-          LD_LIBRARY_PATH: ${{ github.workspace }}/libduckdb
-      - name: Upload to codecov.io
-        if: matrix.os == 'ubuntu-latest'
-        uses: codecov/codecov-action@v1
 
-  sanitizer:
+  Windows:
+    name: Window build from Source
+    needs: test
+    runs-on: windows-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: hecrj/setup-rust-action@v1
+        with:
+          rust-version: stable
+          targets: x86_64-pc-windows-msvc
+      - run: cargo test --features "bundled" --features "modern-full"
+
+  Sanitizer:
     name: Address Sanitizer
     needs: test
     runs-on: ubuntu-latest

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -45,10 +45,17 @@ jobs:
           fileName: ${{ matrix.duckdb }}
           out-file-path: .
       - name: Extract duckdb
+        if: matrix.os == 'ubuntu-latest'
         uses: ihiroky/extract-action@v1
         with:
           file_path: ${{ github.workspace }}/${{ matrix.duckdb }}
           extract_dir: libduckdb
+      - name: Extract duckdb
+        if: matrix.os == 'windows-latest'
+        uses: ihiroky/extract-action@v1
+        with:
+          file_path: ${env:GITHUB_WORKSPACE}\${{ matrix.duckdb }}
+          extract_dir: ${{ github.workspace }}/libduckdb
 
       # - run: cargo clippy --all-targets --workspace --features bundled --features modern-full -- -D warnings -A clippy::redundant-closure
       - run: cargo clippy --all-targets --workspace --features buildtime_bindgen --features modern-full -- -D warnings -A clippy::redundant-closure

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -35,6 +35,7 @@ jobs:
           components: 'rustfmt, clippy'
 
       - run: cargo fmt --all -- --check
+        if: matrix.os == 'ubuntu-latest'
       
       # download libduckdb
       - uses: robinraju/release-downloader@v1.4
@@ -66,13 +67,18 @@ jobs:
           DUCKDB_LIB_DIR: ${{ github.workspace }}/libduckdb
           DUCKDB_INCLUDE_DIR: ${{ github.workspace }}/libduckdb
           LD_LIBRARY_PATH: ${{ github.workspace }}/libduckdb
+      - name: Add path to PATH environment variable
+        if: matrix.os == 'windows-latest'
+        uses: myci-actions/export-env-var-powershell@1
+        with:
+          name: PATH
+          value: $env:PATH;${{ github.workspace }}/libduckdb
       - name: Run cargo-test
         if: matrix.os == 'windows-latest'
         run: cargo test --features "buildtime_bindgen" --features "modern-full"
         env:
           DUCKDB_LIB_DIR: ${{ github.workspace }}/libduckdb
           DUCKDB_INCLUDE_DIR: ${{ github.workspace }}/libduckdb
-          PATH: ${{ github.workspace }}/libduckdb
       - name: Run cargo-tarpaulin
         if: matrix.os == 'ubuntu-latest'
         uses: actions-rs/tarpaulin@v0.1

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -52,7 +52,7 @@ jobs:
           extract_dir: libduckdb
       - name: Windows extract duckdb
         if: matrix.os == 'windows-latest'
-        uses: DuckSoft/extract-7z-action@v1
+        uses: DuckSoft/extract-7z-action@v1.0
         with:
           # file_path: ${{ github.workspace }}\${{ matrix.duckdb }} # not work
           pathSource: D:\a\duckdb-rs\duckdb-rs\${{ matrix.duckdb }}

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -72,7 +72,7 @@ jobs:
         env:
           DUCKDB_LIB_DIR: ${{ github.workspace }}/libduckdb
           DUCKDB_INCLUDE_DIR: ${{ github.workspace }}/libduckdb
-          LD_LIBRARY_PATH: ${{ github.workspace }}/libduckdb
+          PATH: ${{ github.workspace }}/libduckdb
       - name: Run cargo-tarpaulin
         if: matrix.os == 'ubuntu-latest'
         uses: actions-rs/tarpaulin@v0.1

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -95,7 +95,7 @@ jobs:
           DUCKDB_INCLUDE_DIR: ${{ github.workspace }}/libduckdb
 
   Windows:
-    name: Window build from Source
+    name: Windows build from source
     needs: test
     runs-on: windows-latest
     steps:

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -42,11 +42,10 @@ jobs:
           repository: "duckdb/duckdb"
           tag: "v0.4.0"
           fileName: ${{ matrix.duckdb }}
-          rename: "libduckdb.zip"
       - name: Extract duckdb
         uses: ihiroky/extract-action@v1
         with:
-          file_path: libduckdb.zip
+          file_path: ${{ github.workspace }}/${{ matrix.duckdb }}
           extract_dir: libduckdb
       # - name: Download DuckDB
       #   run: |

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -14,7 +14,7 @@ jobs:
       fail-fast: true
       matrix:
         include:
-          #- { target: x86_64-pc-windows-msvc, os: windows-latest }
+          - { target: x86_64-pc-windows-msvc, os: windows-latest }
           - { target: x86_64-unknown-linux-gnu, os: ubuntu-latest }
           #- { target: x86_64-apple-darwin, os: macos-latest }
           #- {

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -54,7 +54,8 @@ jobs:
         if: matrix.os == 'windows-latest'
         uses: ihiroky/extract-action@v1
         with:
-          file_path: ${env:GITHUB_WORKSPACE}\${{ matrix.duckdb }}
+          # file_path: ${{ github.workspace }}\${{ matrix.duckdb }} # not work
+          file_path: D:\a\duckdb-rs\duckdb-rs\${{ matrix.duckdb }}
           extract_dir: ${{ github.workspace }}/libduckdb
 
       # - run: cargo clippy --all-targets --workspace --features bundled --features modern-full -- -D warnings -A clippy::redundant-closure

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -36,21 +36,20 @@ jobs:
 
       - run: cargo fmt --all -- --check
       
-
+      # download libduckdb
       - uses: robinraju/release-downloader@v1.4
+        name: Download duckdb
         with:
           repository: "duckdb/duckdb"
           tag: "v0.4.0"
           fileName: ${{ matrix.duckdb }}
+          out-file-path: .
       - name: Extract duckdb
         uses: ihiroky/extract-action@v1
         with:
           file_path: ${{ github.workspace }}/${{ matrix.duckdb }}
           extract_dir: libduckdb
-      # - name: Download DuckDB
-      #   run: |
-      #     wget https://github.com/duckdb/duckdb/releases/download/v0.4.0/libduckdb-linux-amd64.zip -O libduckdb.zip
-      #     unzip libduckdb.zip -d libduckdb
+
       # - run: cargo clippy --all-targets --workspace --features bundled --features modern-full -- -D warnings -A clippy::redundant-closure
       - run: cargo clippy --all-targets --workspace --features buildtime_bindgen --features modern-full -- -D warnings -A clippy::redundant-closure
         name: run cargo clippy
@@ -70,6 +69,7 @@ jobs:
           DUCKDB_INCLUDE_DIR: ${{ github.workspace }}/libduckdb
           LD_LIBRARY_PATH: ${{ github.workspace }}/libduckdb
       - name: Upload to codecov.io
+        if: matrix.os == 'ubuntu-latest'
         uses: codecov/codecov-action@v1
 
   sanitizer:

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -14,8 +14,8 @@ jobs:
       fail-fast: true
       matrix:
         include:
-          - { target: x86_64-pc-windows-msvc, os: windows-latest }
-          - { target: x86_64-unknown-linux-gnu, os: ubuntu-latest }
+          - { target: x86_64-pc-windows-msvc, os: windows-latest, duckdb: libduckdb-windows-amd64.zip }
+          - { target: x86_64-unknown-linux-gnu, os: ubuntu-latest, duckdb: libduckdb-linux-amd64.zip }
           #- { target: x86_64-apple-darwin, os: macos-latest }
           #- {
               #target: x86_64-pc-windows-gnu,
@@ -35,10 +35,23 @@ jobs:
           components: 'rustfmt, clippy'
 
       - run: cargo fmt --all -- --check
-      - name: Download DuckDB
-        run: |
-          wget https://github.com/duckdb/duckdb/releases/download/v0.4.0/libduckdb-linux-amd64.zip -O libduckdb.zip
-          unzip libduckdb.zip -d libduckdb
+      
+
+      - uses: robinraju/release-downloader@v1.4
+        with:
+          repository: "duckdb/duckdb"
+          tag: "v0.4.0"
+          fileName: ${{ matrix.duckdb }}
+          rename: "libduckdb.zip"
+      - name: Extract duckdb
+        uses: ihiroky/extract-action@v1
+        with:
+          file_path: libduckdb.zip
+          extract_dir: libduckdb
+      # - name: Download DuckDB
+      #   run: |
+      #     wget https://github.com/duckdb/duckdb/releases/download/v0.4.0/libduckdb-linux-amd64.zip -O libduckdb.zip
+      #     unzip libduckdb.zip -d libduckdb
       # - run: cargo clippy --all-targets --workspace --features bundled --features modern-full -- -D warnings -A clippy::redundant-closure
       - run: cargo clippy --all-targets --workspace --features buildtime_bindgen --features modern-full -- -D warnings -A clippy::redundant-closure
         name: run cargo clippy

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -52,7 +52,7 @@ jobs:
           extract_dir: libduckdb
       - name: Windows extract duckdb
         if: matrix.os == 'windows-latest'
-        uses: extract-7z-action@v1
+        uses: DuckSoft/extract-7z-action@v1
         with:
           # file_path: ${{ github.workspace }}\${{ matrix.duckdb }} # not work
           pathSource: D:\a\duckdb-rs\duckdb-rs\${{ matrix.duckdb }}

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -47,7 +47,7 @@ memchr = "2.3"
 uuid = { version = "1.0", optional = true }
 smallvec = "1.6.1"
 cast = { version = "0.3", features = ["std"] }
-arrow = { version = "17", default-features = false, features = ["prettyprint"] }
+arrow = { version = "18", default-features = false, features = ["prettyprint"] }
 rust_decimal = "1.14"
 strum = { version = "0.24", features = ["derive"] }
 r2d2 = { version = "0.8.9", optional = true }

--- a/libduckdb-sys/Cargo.toml
+++ b/libduckdb-sys/Cargo.toml
@@ -28,4 +28,4 @@ cc = { version = "1.0", features = ["parallel"], optional = true }
 vcpkg = { version = "0.2", optional = true }
 
 [dev-dependencies]
-arrow = { version = "17", default-features = false }
+arrow = { version = "18", default-features = false }

--- a/libduckdb-sys/build.rs
+++ b/libduckdb-sys/build.rs
@@ -65,7 +65,6 @@ mod build_bundled {
             .cpp(true)
             // .static_flag(true)
             .shared_flag(true)
-            .pic(true)
             .flag_if_supported("-std=c++11")
             .flag_if_supported("-stdlib=libc++")
             .flag_if_supported("-stdlib=libstdc++")

--- a/libduckdb-sys/build.rs
+++ b/libduckdb-sys/build.rs
@@ -69,6 +69,13 @@ mod build_bundled {
             .flag_if_supported("-stdlib=libc++")
             .flag_if_supported("-stdlib=libstdc++")
             .warnings(false);
+
+        let compiler = cfg.get_compiler();
+
+        if compiler.is_like_msvc() {
+            cfg.flag("/bigobj");
+        }
+
         cfg.compile(lib_name);
 
         println!("cargo:lib_dir={}", out_dir);

--- a/libduckdb-sys/build.rs
+++ b/libduckdb-sys/build.rs
@@ -38,6 +38,8 @@ fn main() {
 mod build_bundled {
     use std::path::Path;
 
+    use crate::win_target;
+
     pub fn main(out_dir: &str, out_path: &Path) {
         let lib_name = super::lib_name();
 
@@ -63,17 +65,14 @@ mod build_bundled {
         let mut cfg = cc::Build::new();
         cfg.file(format!("{}/duckdb.cpp", lib_name))
             .cpp(true)
-            // .static_flag(true)
-            .shared_flag(true)
             .flag_if_supported("-std=c++11")
             .flag_if_supported("-stdlib=libc++")
             .flag_if_supported("-stdlib=libstdc++")
+            .flag_if_supported("/bigobj")
             .warnings(false);
 
-        let compiler = cfg.get_compiler();
-
-        if compiler.is_like_msvc() {
-            cfg.flag("/bigobj");
+        if win_target() {
+            cfg.define("DUCKDB_BUILD_LIBRARY", None);
         }
 
         cfg.compile(lib_name);

--- a/libduckdb-sys/src/error.rs
+++ b/libduckdb-sys/src/error.rs
@@ -1,6 +1,6 @@
+use crate::duckdb_state;
 use std::error;
 use std::fmt;
-use std::os::raw::c_uint;
 
 /// Error Codes
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
@@ -59,11 +59,11 @@ pub enum ErrorCode {
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
 pub struct Error {
     pub code: ErrorCode,
-    pub extended_code: c_uint,
+    pub extended_code: duckdb_state,
 }
 
 impl Error {
-    pub fn new(result_code: c_uint) -> Error {
+    pub fn new(result_code: duckdb_state) -> Error {
         Error {
             code: ErrorCode::Unknown,
             extended_code: result_code,
@@ -88,6 +88,6 @@ impl error::Error for Error {
     }
 }
 
-pub fn code_to_str(_: c_uint) -> &'static str {
+pub fn code_to_str(_: duckdb_state) -> &'static str {
     "Unknown error code"
 }

--- a/libduckdb-sys/src/lib.rs
+++ b/libduckdb-sys/src/lib.rs
@@ -13,8 +13,8 @@ pub use bindings::*;
 
 use std::os::raw::c_uint;
 
-pub const DuckDBError: c_uint = duckdb_state_DuckDBError;
-pub const DuckDBSuccess: c_uint = duckdb_state_DuckDBSuccess;
+pub const DuckDBError: c_uint = duckdb_state_DuckDBError as c_uint;
+pub const DuckDBSuccess: c_uint = duckdb_state_DuckDBSuccess as c_uint;
 
 pub use self::error::*;
 mod error;

--- a/libduckdb-sys/src/lib.rs
+++ b/libduckdb-sys/src/lib.rs
@@ -11,10 +11,8 @@ mod bindings {
 #[allow(clippy::all)]
 pub use bindings::*;
 
-use std::os::raw::c_uint;
-
-pub const DuckDBError: c_uint = duckdb_state_DuckDBError as c_uint;
-pub const DuckDBSuccess: c_uint = duckdb_state_DuckDBSuccess as c_uint;
+pub const DuckDBError: duckdb_state = duckdb_state_DuckDBError;
+pub const DuckDBSuccess: duckdb_state = duckdb_state_DuckDBSuccess;
 
 pub use self::error::*;
 mod error;

--- a/src/config.rs
+++ b/src/config.rs
@@ -116,7 +116,7 @@ impl Config {
         };
         if state != ffi::DuckDBSuccess {
             return Err(Error::DuckDBFailure(
-                ffi::Error::new(state as u32),
+                ffi::Error::new(state),
                 Some(format!("set {}:{} error", key, value)),
             ));
         }

--- a/src/error.rs
+++ b/src/error.rs
@@ -5,7 +5,6 @@ use crate::types::Type;
 use std::error;
 use std::ffi::CStr;
 use std::fmt;
-use std::os::raw::c_uint;
 use std::path::PathBuf;
 use std::str;
 
@@ -216,13 +215,13 @@ impl error::Error for Error {
 // These are public but not re-exported by lib.rs, so only visible within crate.
 
 #[inline]
-fn error_from_duckdb_code(code: c_uint, message: Option<String>) -> Result<()> {
-    Err(Error::DuckDBFailure(ffi::Error::new(code as u32), message))
+fn error_from_duckdb_code(code: ffi::duckdb_state, message: Option<String>) -> Result<()> {
+    Err(Error::DuckDBFailure(ffi::Error::new(code), message))
 }
 
 #[cold]
 #[inline]
-pub fn result_from_duckdb_appender(code: c_uint, mut appender: ffi::duckdb_appender) -> Result<()> {
+pub fn result_from_duckdb_appender(code: ffi::duckdb_state, mut appender: ffi::duckdb_appender) -> Result<()> {
     if code == ffi::DuckDBSuccess {
         return Ok(());
     }
@@ -241,7 +240,7 @@ pub fn result_from_duckdb_appender(code: c_uint, mut appender: ffi::duckdb_appen
 
 #[cold]
 #[inline]
-pub fn result_from_duckdb_prepare(code: c_uint, mut prepare: ffi::duckdb_prepared_statement) -> Result<()> {
+pub fn result_from_duckdb_prepare(code: ffi::duckdb_state, mut prepare: ffi::duckdb_prepared_statement) -> Result<()> {
     if code == ffi::DuckDBSuccess {
         return Ok(());
     }
@@ -260,7 +259,7 @@ pub fn result_from_duckdb_prepare(code: c_uint, mut prepare: ffi::duckdb_prepare
 
 #[cold]
 #[inline]
-pub fn result_from_duckdb_arrow(code: c_uint, mut out: ffi::duckdb_arrow) -> Result<()> {
+pub fn result_from_duckdb_arrow(code: ffi::duckdb_state, mut out: ffi::duckdb_arrow) -> Result<()> {
     if code == ffi::DuckDBSuccess {
         return Ok(());
     }

--- a/src/inner_connection.rs
+++ b/src/inner_connection.rs
@@ -39,7 +39,7 @@ impl InnerConnection {
             if r != ffi::DuckDBSuccess {
                 let msg = Some(CStr::from_ptr(c_err).to_string_lossy().to_string());
                 ffi::duckdb_free(c_err as *mut c_void);
-                return Err(Error::DuckDBFailure(ffi::Error::new(r as u32), msg));
+                return Err(Error::DuckDBFailure(ffi::Error::new(r), msg));
             }
             InnerConnection::new(db, true)
         }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -46,7 +46,7 @@
 //!     for person in person_iter {
 //!         println!("Found person {:?}", person.unwrap());
 //!     }
-//!    
+//!
 //!     // query table by arrow
 //!     let rbs: Vec<RecordBatch> = stmt.query_arrow([])?.collect();
 //!     print_batches(&rbs);
@@ -558,6 +558,7 @@ mod test {
     }
 
     #[test]
+    #[cfg_attr(windows, ignore = "Windows doesn't allow concurrent writes to a file")]
     fn test_concurrent_transactions_busy_commit() -> Result<()> {
         let tmp = tempfile::tempdir().unwrap();
         let path = tmp.path().join("transactions.db3");

--- a/src/row.rs
+++ b/src/row.rs
@@ -4,7 +4,7 @@ use std::sync::Arc;
 use super::{Error, Result, Statement};
 use crate::types::{self, FromSql, FromSqlError, ValueRef};
 
-use arrow::array::{self, Array, StructArray};
+use arrow::array::{self, Array, BasicDecimalArray, StructArray};
 use arrow::datatypes::*;
 use fallible_iterator::FallibleIterator;
 use fallible_streaming_iterator::FallibleStreamingIterator;

--- a/src/statement.rs
+++ b/src/statement.rs
@@ -624,6 +624,7 @@ mod test {
     }
 
     #[test]
+    #[cfg_attr(windows, ignore = "Windows doesn't allow concurrent writes to a file")]
     fn test_insert_duplicate() -> Result<()> {
         let db = Connection::open_in_memory()?;
         db.execute_batch("CREATE TABLE foo(x INTEGER UNIQUE)")?;

